ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -622,6 +622,24 @@ static const upb_pbdecodermethod *msgdef_decodermethod(Descriptor* desc) {
   return desc->fill_method;
 }
 
+// Closure of information required within an encode/decode-error callback: the
+// environment to free, and the error message template to use when raising a
+// Ruby exception.
+typedef struct {
+  upb_env* env;
+  const char* msg;
+} env_errdata;
+
+// Callback invoked by upb if any error occurs during parsing or serialization.
+static bool env_error_func(void* ud, const upb_status* status) {
+  env_errdata* errdata = ud;
+  // Free the env -- rb_raise will longjmp up the stack past the encode/decode
+  // function so it would not otherwise have been freed.
+  upb_env_uninit(errdata->env);
```

Comment:
> I think you're just thinking through upb semantics here

Yes exactly, you're doing something that no one has ever done before, which I've thought about a little, but not deeply. :)